hv.c

```diff
@@ -1460,30 +1451,6 @@ S_hsplit(pTHX_ HV *hv, STRLEN const oldsize, STRLEN newsize)
 #endif
     HvARRAY(hv) = (HE**) a;
     HvMAX(hv) = newsize - 1;
-    /* before we zero the newly added memory, we
-     * need to deal with the aux struct that may be there
-     * or have been allocated by us*/
-    if (do_aux) {
-        struct xpvhv_aux *const dest = HvAUX(hv);
-        if (SvOOK(hv)) {
-            /* we reset the iterator's xhv_rand as well, so they get a totally new ordering */
-#ifdef PERL_HASH_RANDOMIZE_KEYS
-            dest->xhv_rand = (U32)PL_hash_rand_bits;
-#endif
```

Comment:
If I read things correctly, this was already a noop (since we've already set it on insert before deciding we need a split). If that's correct, it might be worth removing these 3 lines as a separate commit; if it's not, it probably needs mentioning in the commit message that we now no longer re-randomize on split.

Reply:
> If I read things correctly

Oh my. Yes, I've been going backwards and forwards for a while trying to be sure about this.

I think that you are correct that it's a no-op for an insert. (Well spotted. I'd not even noticed that this might be relevant.)

But. It doesn't seem to be a no-op where `S_hsplit` is called by `Perl_hv_ksplit`, which is `keys` as an LVALUE where the hash is expanded. So things like `perl -MDevel::Peek -lwe '%h = (1..4); Dump \%h; print scalar each %h; Dump \%h; keys %h = 20; Dump \%h; print scalar each %h; Dump \%h'`

In blead that resets the iterator *and* forces a new random ordering. In this branch, as a result of the commit above it doesn't. (The line `RAND = 0x......` in the rather verbose debug output)

I suspect that the correct thing is to split those 3 lines out as a separate commit that also inserts them into `Perl_hv_ksplit`, along with comments explaining what's going on, but I want to sleep on this idea.

Reply:
> It doesn't seem to be a no-op where `S_hsplit` is called by `Perl_hv_ksplit`

Ah right, I hadn't considered lvalue keys().

> I suspect that the correct thing is to split those 3 lines out as a separate commit that also inserts them into `Perl_hv_ksplit`, along with comments explaining what's going on, but I want to sleep on this idea.

That makes sense, if seed reset is really required for this case. It isn't obvious to me that it _is_ required (other than for continuity with current behaviour) for a request to change the number of buckets that doesn't change the actual list of keys.

Reply:
> That makes sense, if seed reset is really required for this case. It isn't obvious to me that it _is_ required (other than for continuity with current behaviour) for a request to change the number of buckets that doesn't change the actual list of keys.

Agree with your reasoning. But it was easy enough to keep the current behaviour, so (for now) I did, as e9dd222847f9